File: modules/eventing/sink_format_test.go
// SPDX-FileCopyrightText: 2026 Olivares.AI
// SPDX-License-Identifier: AGPL-3.0-only
// Additional terms under AGPL-3.0-only section 7(a) disclaim warranty and limit liability: see DISCLAIMER.md at the repository root.
package eventing

import (
	"sort"
	"testing"

	"github.com/olivaresai/olivares/core/audit"
	"github.com/olivaresai/olivares/sdk/siemwire"
)

// TestSinkFormatsAreTheCatalogEventingSubset: the allow-list DERIVES from the
// sdk/siemwire catalog, and this test pins the relationship in both directions
// plus the one deliberate asymmetry against the ledger registry. Before the
// catalog, this test asserted "eventing == ledger registry + json"; the catalog
// remap made that relationship subset MATH rather than a coincidence of two
// hand-maintained lists: eventing == (ledger − otlp_log_record) + json.
func TestSinkFormatsAreTheCatalogEventingSubset(t *testing.T) {
	set := siemwire.EventingSinkFormats()
	for _, f := range set.Tokens() {
		if !validSinkFormat(string(f)) {
			t.Errorf("sink rejects %q, which the catalog's eventing subset declares", f)
		}
	}
	if got, want := len(sinkFormats), len(set.Tokens()); got != want {
		t.Errorf("sinkFormats has %d entries, want %d from the catalog; got %v",
			got, want, sinkFormatKeys())
	}
	if validSinkFormat("not-a-format") {
		t.Error("the allow-list must stay closed")
	}
	// "" means "use the per-sink default" and must stay allowed; the default the
	// engine applies is the catalog's surface default.
	if !validSinkFormat("") {
		t.Error("an unset format must fall back to the per-sink default")
	}
	if set.Default() != siemwire.TokenOCSF {
		// OCSF has been the SIEM-sink default since the sink profile shipped.
		t.Errorf("eventing surface default = %q, want ocsf", set.Default())
	}

	// The deliberate asymmetry against the ledger registry, stated as math:
	// every ledger format EXCEPT otlp_log_record is declarable (a sink POSTs one
	// rendered body per event and a bare LogRecord line is not an OTLP /v1/logs
	// request), and json is declarable ONLY here (the structured passthrough
	// modules/siemforward intercepts before any audit encoder runs).
	for _, f := range audit.Formats() {
		declarable := validSinkFormat(string(f))
		if f == audit.FormatOTLPLogRecord {
			if declarable {
				t.Errorf("sink accepts %q — the bare projection is a pull-export capability, excluded by design", f)
			}
			continue
		}
		if !declarable {
			t.Errorf("sink rejects %q, which the ledger export can produce", f)
		}
	}
	if !validSinkFormat("json") {
		t.Error("json (the structured passthrough) must stay declarable")
	}
	if audit.ValidFormat("json") {
		t.Error("json must NOT be a ledger export format — it exists only on the eventing surface")
	}
}

// sinkFormatKeys renders the allow-list in sorted order so a failure message reads
// the same on every run (map iteration order is randomized).
func sinkFormatKeys() []string {
	keys := make([]string, 0, len(sinkFormats))
	for f := range sinkFormats {
		keys = append(keys, f)
	}
	sort.Strings(keys)
	return keys
}

File: web/src/features/eventing/sink-format.test.ts
// SPDX-FileCopyrightText: 2026 Olivares.AI
// SPDX-License-Identifier: AGPL-3.0-only
// Additional terms under AGPL-3.0-only section 7(a) disclaim warranty and limit liability: see DISCLAIMER.md at the repository root.
//
// The subscription form's sink_format list is a closed set mirroring the
// engine's catalog. Like the audit export-format pin, the comparison is against
// the COMMITTED beta OpenAPI snapshot rather than a second TypeScript literal:
// the snapshot's enum is rendered by the engine from the sdk/siemwire catalog's
// eventing surface — the same registry the eventing module validates against —
// and `task openapi:check` fails on generator/snapshot drift. A format added to
// the catalog and forgotten here fails on this side; one invented here that the
// engine refuses fails too.
import { describe, expect, it } from 'vitest'
import openapiBeta from '../../../openapi/openapi.beta.json'
import { SINK_FORMATS } from './eventing-view'

/** The engine-owned sink_format enum on POST /v1/m/eventing/subscriptions. */
function engineSinkFormats(): string[] {
  const doc = openapiBeta as unknown as {
    paths: Record<
      string,
      {
        post: {
          requestBody?: {
            content: Record<
              string,
              {
                schema: {
                  properties?: Record<string, { enum?: string[] }>
                }
              }
            >
          }
        }
      }
    >
  }
  const body = doc.paths['/v1/m/eventing/subscriptions']?.post?.requestBody
  if (!body) {
    throw new Error(
      'the beta OpenAPI snapshot no longer declares the subscription request body',
    )
  }
  const enumValues =
    body.content['application/json']?.schema.properties?.['sink_format']?.enum
  if (!enumValues?.length) {
    throw new Error('the beta snapshot has no enum for sink_format')
  }
  return enumValues
}

describe('eventing sink formats', () => {
  it('offers exactly the formats the engine accepts, in the catalog order', () => {
    // The engine's enum leads with '' (unset = the surface default); the form
    // models that as the absence of a selection, so the mirror is the enum
    // minus the empty spelling, order preserved.
    const engine = engineSinkFormats()
    expect(engine[0]).toBe('')
    expect([...SINK_FORMATS]).toEqual(engine.slice(1))
  })
})
